plugins/reachability/src/main/java/org/qbicc/plugin/reachability/ReachabilityBlockBuilder.java

```diff
@@ -202,150 +199,162 @@ public Value classOf(Value typeId) {
+        if (!info.isInstantiatedClass(type)) {
             if (directlyInstantiated) {
                 rtaLog.debugf("Adding class %s (instantiated in %s)", type.getDescriptor().getClassName(), originalElement);
-            } else if (arrayElement) {
-                rtaLog.debugf("Adding class %s (array element in %s)", type.getDescriptor().getClassName(), originalElement);
             } else {
                 rtaLog.debugf("\tadding ancestor class: %s", type.getDescriptor().getClassName());
             }
+            info.addReachableClass(type);
+            info.addInstantiatedClass(type);
+
             if (type.hasSuperClass()) {
-                processInstantiatedClass(type.getSuperClass(), false, false);
-                info.addLiveClass(type);
+                processInstantiatedClass(type.getSuperClass(), false);
 
-                // For every instance method that is not already enqueued,
-                // check to see if it is overriding an enqueued method and thus should be enqueued.
+                // For every instance method that is not already invokable,
+                // check to see if it is either (a) a deferred invoked method or (b) overriding an invokable method and thus should be enqueued.
                 for (MethodElement im: type.getInstanceMethods()) {
-                    if (!ctxt.wasEnqueued(im)) {
-                        MethodElement overiddenMethod = type.getSuperClass().resolveMethodElementVirtual(im.getName(), im.getDescriptor());
-                        if (overiddenMethod != null && ctxt.wasEnqueued(overiddenMethod)) {
+                    if (!info.isInvokableMethod(im)) {
+                        if (info.isDeferredInstanceMethod(im)) {
+                            rtaLog.debugf("\tnewly reachable class: enqueued deferred instance method: %s", im);
+                            info.removeDeferredInstanceMethod(im);
+                            info.addInvokableMethod(im);
                             ctxt.enqueue(im);
-                            rtaLog.debugf("\tnewly reachable class: enqueued overriding method: %s", im);
+                        } else {
+                            MethodElement overiddenMethod = type.getSuperClass().resolveMethodElementVirtual(im.getName(), im.getDescriptor());
+                            if (overiddenMethod != null && info.isInvokableMethod(overiddenMethod)) {
+                                rtaLog.debugf("\tnewly reachable class: enqueued overriding instance method: %s", im);
+                                info.addInvokableMethod(im);
+                                ctxt.enqueue(im);
+                            }
                         }
                     }
                 }
             }
 
             // Extend the interface hierarchy
             for (LoadedTypeDefinition i: type.getInterfaces()) {
-                processInstantiatedInterface(info, i, false);
+                processInstantiatedInterface(i);
```

Comment:
`type.getInterfaces()` returns the locally declared interfaces.  Does this need to walk the hierarchy to find all the interfaces this class can process?

Ah, I see now.  The solution is recursively built up so any interfaces declared on the super classes are handled by the call to `processInstantiatedClass` passing the superclass